docs/configuration.md
# Configuration
Configuration is done through environment variables. The following subsections 
detail how to configure BlockMe to work in your environment.

## Logging
Logging configurations allow users to define how and where log messages are 
output

**LOG_FORMAT** | `text_pretty`, `text`, `json` | Default: `text_pretty`
> Defines the format to output requests logs in. text_pretty is for console 
> logging only and outputs request logs in a nice human-readable format. 'text' 
> and 'json' output log data into a format that is better for automated log 
> collection

**LOG_LEVEL** | `DEBUG`, `INFO`, `WARNING`, `ERROR` | Default: `INFO`
> Set the level of log verbosity

**LOG_LOCATION** | `console` or `file` | Default: `console`
> Set the location to output log files to. 'console' outputs to the standard 
> output, and 'file' outputs to ./blockme.log. NOTE: Setting LOG_FORMAT to 
> 'text_pretty' will always output to the standards output (console)

## Database
Setting up a database enabled persistent IP storage to ensure that blocked IPs 
stay blocked.

**DB_TYPE** | `postgres` or `sqlite` | Default: `sqlite`  
> Choosing Postgres makes DB_HOST, DB_PORT, DB_NAME, DB_USER, & DB_PASS required 
> in order to connect to your postgres database. Sqlite database files are 
> stored the database subdirectory of your working directory (`/app/database/` 
> for Docker containers)

**DB_HOST** | IP address or FQDN | Default: N/A
> Routable address of your database server. The DB_HOST environment variable has 
> no default and will result in the program exiting with an error code if the 
> DB_TYPE has been set to postgres and no DB_HOST is specified.

**DB_PORT** | 1 - 65535 | Default: `sqlite`
> TCP port that your database server is listening on. This variable will default
> to the default port of the given DB_TYPE (Postgres → 5432, MariaDB → 3306).

**DB_NAME** | Existing database name | Default: `blockme`
> Specify a database name if you want to use a specific database name. If no 
> database exists with the given name, BlockMe will attempt to create it.

**DB_USER** | Existing database user | Default: None
> The DB_USER environment variable has no default and will result in the program 
> exiting with an error code if the DB_TYPE has been set to postgres and no 
> DB_USER is specified.

**DB_PASS** | Existing database user password | Default: None
> The DB_PASS environment variable has no default and will result in the program
> exiting with an error code if the DB_TYPE has been set to postgres and no
> DB_PASS is specified.

## Reset
The reset functionality allows user's to unblock themselves with a generated 
UUID.

**RESET_ME** | `true` or `false` | Default: `false`
> Setting RESET_ME to `true` enables the reset functionality.

**RESET_ME_CRON** | Five field Cron schedule | Default: `1 * * * *`
> Defines how often the reset token is rotated and sent to the webhook.

**RESET_ME_WEBHOOK** | Existing webhook URL (Discord) | Default: None
> Defines where reset tokens will be sent to.

## Misc

**IGNORE_LIST** | Comma separated list of IPs | Default: ""
> Adding an IP address to the IGNORE_LIST environment variable prevents that IP 
> from being blocked. All other logic works the same, and BlockMe will still 
> return a "blocked" message but the address will not be added to the block list

config/config.go
package config

import (
	"errors"
	"fmt"
	"net"
	"os"
	"regexp"
	"strconv"
	"strings"

	"github.com/joho/godotenv"
)

type EnvConfig struct {
	LogFormat   string
	LogLevel    string
	LogLocation string

	DbType string
	DbHost string
	DbPort int
	DbName string
	DbUser string
	DbPass string

	ResetMe        bool
	ResetMeCron    string
	ResetMeWebhook string

	IgnoreList []string
}

var Env *EnvConfig
var ErrMissingRequiredEnvVar = errors.New("missing required config")

func InitConfig() error {
	fmt.Printf("\nInitializing config...\n")
	_ = godotenv.Load()

	fmt.Println("config: Initializing Logging configs")
	logFormat := os.Getenv("LOG_FORMAT")
	if logFormat != "text_pretty" && logFormat != "text" && logFormat != "json" {
		fmt.Println("config: Invalid LOG_FORMAT - Continuing with text_pretty")
		logFormat = "text_pretty"
	}
	logLevel := os.Getenv("LOG_LEVEL")
	if logLevel != "DEBUG" && logLevel != "INFO" && logLevel != "WARNING" && logLevel != "ERROR" {
		fmt.Println("config: Invalid LOG_LEVEL - Continuing with INFO")
		logLevel = "INFO"
	}
	logLocation := os.Getenv("LOG_LOCATION")
	if logLocation != "console" && logLocation != "file" {
		fmt.Println("config: Invalid LOG_LOCATION - Continuing with console")
		logLocation = "console"
	}

	fmt.Printf("config: Initializing DB values\n")
	dbType := strings.ToLower(os.Getenv("DB_TYPE"))
	if dbType != "postgres" && dbType != "sqlite" {
		fmt.Println("config: DB_TYPE invalid or not set - Defaulting to sqlite")
		dbType = "sqlite"
	}
	dbHost := os.Getenv("DB_HOST")
	if dbHost == "" && dbType == "postgres" {
		return fmt.Errorf("%w: %s", ErrMissingRequiredEnvVar, "DB_HOST")
	}
	dbPortStr := os.Getenv("DB_PORT")
	if dbPortStr == "" && dbType == "postgres" {
		dbPortStr = "0"
	}
	dbPort, _ := strconv.Atoi(dbPortStr)
	if dbPort <= 0 || dbPort > 65535 {
		switch dbType {
		case "postgres":
			fmt.Println("config: Invalid DB_PORT - Continuing with Postgres default 5432")
			dbPort = 5432
			break
		default:
			dbPort = 0
		}
	}
	dbName := os.Getenv("DB_NAME")
	if dbName == "" && dbType == "postgres" {
		fmt.Println("config: Invalid or missing DB_NAME - Continuing with default 'blockme'")
		dbName = "blockme"
	}
	dbUser := os.Getenv("DB_USER")
	if dbUser == "" && dbType == "postgres" {
		return fmt.Errorf("%w: %s", ErrMissingRequiredEnvVar, "DB_USER")
	}
	dbPass := os.Getenv("DB_PASS")
	if dbPass == "" && dbType == "postgres" {
		return fmt.Errorf("%w: %s", ErrMissingRequiredEnvVar, "DB_PASS")
	}

	fmt.Printf("config: Initializing reset values\n")
	resetMeStr := os.Getenv("RESET_ME")
	resetMe, err := strconv.ParseBool(resetMeStr)
	if err != nil {
		fmt.Println("config: Invalid or missing RESET_ME - Continuing with default false")
		resetMe = false
	}
	var (
		resetMeCron    string
		resetMeWebhook string
	)

	if resetMe {
		resetMeCron = os.Getenv("RESET_ME_CRON")
		if resetMeCron == "" {
			fmt.Println("config: Invalid or missing RESET_ME_CRON - Continuing with default '1 * * * *")
			resetMeCron = "1 * * * *"
		}
		resetMeWebhook = os.Getenv("RESET_ME_WEBHOOK")
		if resetMeWebhook == "" {
			fmt.Println("config: Missing RESET_ME_WEBHOOK while RESET_ME is enabled - Disabling RESET_ME")
			resetMe = false
		}
	}

	// Remove all chars not in the regex, split on commas, and parse each list item as an IP
	re := regexp.MustCompile("[^0-9.,]")
	ignoreListStr := os.Getenv("IGNORE_LIST")
	ignoreListStr = re.ReplaceAllString(ignoreListStr, "")
	var ignoreList []string
	for _, ignore := range strings.Split(ignoreListStr, ",") {
		addr := net.ParseIP(ignore)
		if addr != nil {
			ignoreList = append(ignoreList, addr.String())
		}
	}

	config := EnvConfig{
		LogFormat:   logFormat,
		LogLevel:    logLevel,
		LogLocation: logLocation,

		DbType: dbType,
		DbHost: dbHost,
		DbPort: dbPort,
		DbName: dbName,
		DbUser: dbUser,
		DbPass: dbPass,

		ResetMe:        resetMe,
		ResetMeCron:    resetMeCron,
		ResetMeWebhook: resetMeWebhook,

		IgnoreList: ignoreList,
	}

	Env = &config

	fmt.Printf("config: Successfully setup config\n")
	return nil
}
